include/core/CMemory.h

```diff
@@ -1009,6 +1035,37 @@ class CORE_EXPORT CMemoryDebug : private CNonInstantiatable {
         }
     }
 
+    //! Overload for boost::multi_index::multi_index_container.
+    template<typename T, typename I, typename A>
+    static void dynamicSize(const char* name,
+                            const boost::multi_index::multi_index_container<T, I, A>& t,
+                            CMemoryUsage::TMemoryUsagePtr mem) {
+        // It's tricky to determine the container overhead of a multi-index
+        // container.  It can have an arbitrary number of indices, each of which
+        // can be of a different type.  To accurately determine the overhead
+        // would require some serious template metaprogramming to interpret the
+        // "typename I" template argument, and it's just not worth it given the
+        // infrequent and relatively simple usage (generally just two indices
+        // in our current codebase).  Therefore there's an approximation here
+        // that the overhead is 2 pointers per entry per index.
+        using TMultiIndex = boost::multi_index::multi_index_container<T, I, A>;
+        constexpr std::size_t indexCount{
+            boost::mpl::size<typename TMultiIndex::index_type_list>::value};
+        std::string componentName(name);
+
+        std::size_t items = t.size();
+        CMemoryUsage::SMemoryUsage usage(
+            componentName + "::" + typeid(T).name(),
+            items * (sizeof(T) + 2 * indexCount * sizeof(std::size_t)));
+        CMemoryUsage::TMemoryUsagePtr ptr = mem->addChild();
+        ptr->setName(usage);
+
+        componentName += "_item";
+        for (auto i = t.begin(); i != t.end(); ++i) {
```

Comment:
`const auto& i` ?

Reply:
The idiomatic way to iterate with iterators is `auto` rather than `const auto&`.  All the equivalent loops in the file are done like this one.  The `auto` variable is the iterator, not the container element.

But it did make me wonder why all these loops are using iterators rather than `for (const auto& i : t) {`.  I tried changing them and got this warning:

```
/Users/dave/ml-cpp/include/core/CMemory.h:334:30: warning: loop variable 'i' is always a copy because the range of type 'const std::vector<bool>' does not return a reference [-Wrange-loop-analysis]
            for (const auto& i : t) {
```

So I think it's because _sometimes_ the `vector` method will have to deal with `vector<bool>`, so that needs the explicit iterator version to avoid the warning, and all the others are done the same way for consistency.

Obviously I could change the loops _except_ the `vector` one, but I do think with this file that there's a benefit in keeping as much as possible consistent between the methods that deal with different types of container.  There are already subtle differences just dealing with the container structure differences.  I'm not sure it's good to introduce a difference in the way looping is done too.

Reply:
Weird.. Thanks for looking in to it!